src/psyclone/psyir/transformations/debug_checksum_trans.py

```diff
@@ -0,0 +1,203 @@
+
+    >>> psyir = FortranReader().psyir_from_source("""
+    ...     subroutine mysubroutine()
+    ...     integer, dimension(10,10) :: A
+    ...     integer :: i
+    ...     integer :: j
+    ...     do i = 1, 10
+    ...       do j = 1, 10
+    ...         A(i,j) = A(i,k) + i-j
+    ...       end do
+    ...     end do
+    ...     end subroutine
+    ...     """)
+    ... loop = psyir.children[0].children[0]
+    ... DebugChecksumTrans().apply(loop)
+    ... print(FortranWriter()(psyir))
+    subroutine mysubroutine()
+      integer, dimension(10,10) :: a
+      integer :: i
+      integer :: j
+      integer :: PSYCLONE_INTERNAL_line_
+    <BLANKLINE>
+      do i = 1, 10, 1
+        do j = 1, 10, 1
+          a(i,j) = a(i,j) + i - j
+        enddo
+      enddo
+      PSYCLONE_INTERNAL_line_ = __LINE__
+      PRINT *, "PSyclone checksums from mysubroutine at line:", \
+PSYCLONE_INTERNAL_line_ + 1
+      PRINT *, "a checksum", SUM(a)
+    <BLANKLINE>
+    end subroutine mysubroutine
+    <BLANKLINE>
+
+    '''
+    def apply(self, node: Union[Node, List[Node]], options=None) -> None:
+        '''
+        Applies the checksum transformation to the provided node(s).
+
+        :param node: The node or list of nodes to apply the
+                      transformation to.
+        :param options: a dictionary with options for transformations.
+        :type options: Optional[Dict[str, Any]]
+
+        '''
+        self.validate(node, options={"node-type-check": False})
+
+        node_list = self.get_node_list(node)
+        routine = node_list[0].ancestor(Routine)
+        routine_table = routine.symbol_table
+
+        fwriter = FortranWriter()
+        writes = []
+        # Loop over the assignments in the region
+        assigns = []
+        for node in node_list:
+            assigns.extend(node.walk(Assignment))
+        # Loop through the assignments and find the arrays
+        for assign in assigns:
+            if isinstance(assign.lhs, StructureAccessorMixin):
+                # Find the last member.
+                member = assign.lhs.member
+                while isinstance(member, StructureAccessorMixin):
+                    member = member.member
+                datatype = assign.lhs.datatype
+                while not isinstance(datatype, ScalarType):
+                    datatype = datatype.datatype
+                if (member.is_array and datatype.intrinsic in
+                        [ScalarType.Intrinsic.REAL,
+                         ScalarType.Intrinsic.INTEGER]):
+                    writes.append(assign.lhs)
+            elif (assign.lhs.is_array and assign.lhs.datatype.intrinsic in
+                  [ScalarType.Intrinsic.REAL, ScalarType.Intrinsic.INTEGER]):
+                writes.append(assign.lhs)
+
+        # For each write, add a checksum after.
+        checksum_nodes = []
+        freader = FortranReader()
+        for lhs in writes:
+            # Need to convert the lhs to a full range variant.
```

Comment:
The existing data-extraction code must also do this so may have functionality you can re-use. I think I also had to do it for inlining but that's probably buried in the InlineTrans. In fact, I've had a look and it is very similar but probably not similar enough to be useful.

Reply:
I think the code that does this here is applying to to specific sections and the code that actual does the replacing is very minimal, i.e. just:
```
            # Need to convert the lhs to a full range variant.
            for i in range(len(array_bit.indices)):
                new_index = array_bit.get_full_range(i)
                array_bit.indices[i].replace_with(new_index)
```

I extracted that part out to work on this array_bit instead of being repeated.